src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -597,21 +452,44 @@ private ValueTask WriteSingleChunk<TIOAdapter>(TIOAdapter writeAdapter, ReadOnly
             if (t.IsCompletedSuccessfully)
             {
                 ArrayPool<byte>.Shared.Return(rentedBuffer);
-                FinishWrite();
                 return t;
             }
             else
             {
-                return CompleteAsync(t, rentedBuffer);
+                return CompleteWriteAsync(t, rentedBuffer);
             }
 
-            async ValueTask WaitForWriteIOSlot(TIOAdapter wAdapter, Task lockTask, ReadOnlyMemory<byte> buff)
+            async ValueTask WaitAndWriteAsync(TIOAdapter writeAdapter, ReadOnlyMemory<byte> buffer, Task waitTask, byte[] rentedBuffer)
             {
-                await lockTask.ConfigureAwait(false);
-                await WriteSingleChunk(wAdapter, buff).ConfigureAwait(false);
+                byte[] outBuffer = rentedBuffer;
+                try
+                {
+                    // Wait for renegotiation to finish.
+                    await waitTask.ConfigureAwait(false);
+
+                    SecurityStatusPal status = EncryptData(buffer, ref outBuffer, out int encryptedBytes);
+                    if (status.ErrorCode == SecurityStatusPalErrorCode.TryAgain)
+                    {
+                        // Call WriteSingleChunk() recursively to avoid code duplication.
+                        // This should be extremely rare in cases when second renegotiation happens concurrently with Write.
+                        await WriteSingleChunk(writeAdapter, buffer);
```

Comment:
```suggestion
                        await WriteSingleChunk(writeAdapter, buffer).ConfigureAwait(false);
```

Reply:
We're still holding on to the rentedBuffer at this point.  Is it rare enough that that's not going to matter?

Reply:
I was thinking about but nullability give me some grief. I think it would be rare but since you mentioned it, I made update to return the buffer before WriteSingleChunk() is awaited. 
I was also originally looking for way how to loop it back so we do not need to rent new buffer but that goes against the simplicity we discussed.